File: golang/main_test.go
package main

import (
	"bytes"
	"strings"
	"testing"
	"unicode/utf8"
)

func countWords(s string) int {
	words := strings.Fields(s)
	return len(words)
}

func countLines(s string) int {
	lines := strings.Count(s, "\n")
	if len(s) > 0 && !strings.HasSuffix(s, "\n") {
		lines++
	}
	return lines
}

func countBytes(s string) int {
	return len(s)
}

func countRunes(s string) int {
	return utf8.RuneCountInString(s)
}

func TestCountWords(t *testing.T) {
	// Arrange
	words := "hello world\n"
	expected := countWords(words)

	// Act
	b := bytes.NewBufferString(words)
	result := count(b, false, false, false)

	// Assert
	if result != expected {
		t.Errorf("%v: expected %v got %v instead", words, expected, result)
	}
}

func TestCountLines(t *testing.T) {
	// Arrange
	words := "hello world\n"
	expected := countLines(words)

	// Act
	b := bytes.NewBufferString(words)
	result := count(b, true, false, false)

	// Assert
	if result != expected {
		t.Errorf("%v: expected %v got %v instead", words, expected, result)
	}
}

func TestCountBytes(t *testing.T) {
	// Arrange
	s := "We♥Go" // 7
	expected := countBytes(s)

	// Act
	b := bytes.NewBufferString(s)
	result := count(b, false, true, false)

	// Assert
	if result != expected {
		t.Errorf("%v: expected %v got %v", s, expected, result)
	}
}

func TestCountRunes(t *testing.T) {
	// Arrange
	s := "We♥Go" // 5
	expected := countRunes(s)

	// Act
	b := bytes.NewBufferString(s)
	result := count(b, false, false, true)

	// Assert
	if result != expected {
		t.Errorf("%v: expected %v got %v", s, expected, result)
	}
}

File: denojs/main_test.ts
import { assertEquals } from 'https://deno.land/std@0.170.0/testing/asserts.ts';
import { Buffer } from 'https://deno.land/std@0.170.0/io/buffer.ts';
import { BufReader } from 'https://deno.land/std@0.170.0/io/buf_reader.ts';
import { count } from './main.ts';

Deno.test('count lines in input text', async () => {
  // Arrange
  const input = 'line 1\nline 2\nline 3\n';
  const config = { countLines: true, countBytes: false, countRunes: false };
  const expected = 3;

  // Act
  const buffer = new Buffer(new TextEncoder().encode(input));
  const reader = new BufReader(buffer);
  const result = await count(reader, config);

  // Assert
  assertEquals(result, expected);
});

Deno.test('count bytes in input text', async () => {
  // Arrange
  const input = 'Hello 世界!';
  const config = { countLines: false, countBytes: true, countRunes: false };
  const expected = 13;

  // Act
  const buffer = new Buffer(new TextEncoder().encode(input));
  const reader = new BufReader(buffer);
  const result = await count(reader, config);

  // Assert
  assertEquals(result, expected);
});

Deno.test('count chars in input text', async () => {
  // Arrange
  const input = 'Hello 世界!';
  const config = { countLines: false, countBytes: false, countRunes: true };
  const expected = 9;

  // Act
  const buffer = new Buffer(new TextEncoder().encode(input));
  const reader = new BufReader(buffer);
  const result = await count(reader, config);

  // Assert
  assertEquals(result, expected);
});

Deno.test('count words in input text', async () => {
  // Arrange
  const input = 'Hello 世界!';
  const config = { countLines: false, countBytes: false, countRunes: false };
  const expected = 2;

  // Act
  const buffer = new Buffer(new TextEncoder().encode(input));
  const reader = new BufReader(buffer);
  const result = await count(reader, config);

  // Assert
  assertEquals(result, expected);
});
